openssl-dynamic/src/main/c/sslcontext.c

```diff
@@ -2647,6 +2666,88 @@ TCN_IMPLEMENT_CALL(jboolean, SSLContext, setCurvesList0)(TCN_STDARGS, jlong ctx,
+    if (algorithm == NULL) {
+        tcn_ThrowIllegalArgumentException(e, "Compression algorithm may not be null");
+    }
+    if (!(direction & SSL_CERT_COMPRESSION_DIRECTION_COMPRESS) && !(direction & SSL_CERT_COMPRESSION_DIRECTION_DECOMPRESS)) {
+        tcn_ThrowIllegalArgumentException(e, "Invalid direction specified");
+    }
+
+#ifdef OPENSSL_IS_BORINGSSL
+
+    jclass algorithmClass = (*e)->GetObjectClass(e, algorithm);
+    if (algorithmClass == NULL) {
+        tcn_Throw(e, "Unable to retrieve cert compression algorithm class");
+        return 0;
+    }
+
+    jmethodID compressMethod = (*e)->GetMethodID(e, algorithmClass, "compress", "(J[B)[B");
+    if (compressMethod == NULL) {
+        tcn_ThrowIllegalArgumentException(e, "Unable to retrieve compress method");
+        return 0;
+    }
+
+    jmethodID decompressMethod = (*e)->GetMethodID(e, algorithmClass, "decompress", "(JI[B)[B");
+    if (decompressMethod == NULL) {
+        tcn_ThrowIllegalArgumentException(e, "Unable to retrieve decompress method");
+        return 0;
+    }
+
+    jobject algoRef = (*e)->NewGlobalRef(e, algorithm);
+    if (algoRef == NULL) {
+        tcn_throwOutOfMemoryError(e, "Unable to allocate memory for global cert compression algorithm reference");
+        return 0;
+    }
+
+    int result = 0;
+    switch (algorithmId) {
+        case TLSEXT_cert_compression_zlib:
+            result = SSL_CTX_add_cert_compression_alg(c->ctx, algorithmId,
+                direction & SSL_CERT_COMPRESSION_DIRECTION_COMPRESS ? zlib_compress_java : NULL,
+                direction & SSL_CERT_COMPRESSION_DIRECTION_DECOMPRESS ? zlib_decompress_java : NULL);
+            if (result) {
+                c->ssl_cert_compression_zlib_algorithm = algoRef;
+                c->ssl_cert_compression_zlib_compress_method = compressMethod;
+                c->ssl_cert_compression_zlib_decompress_method = decompressMethod;
+            }
+            break;
+        case TLSEXT_cert_compression_brotli:
+            result = SSL_CTX_add_cert_compression_alg(c->ctx, algorithmId,
+                direction & SSL_CERT_COMPRESSION_DIRECTION_COMPRESS ? brotli_compress_java : NULL,
+                direction & SSL_CERT_COMPRESSION_DIRECTION_DECOMPRESS ? brotli_decompress_java : NULL);
+            if (result) {
+                c->ssl_cert_compression_brotli_algorithm = algoRef;
+                c->ssl_cert_compression_brotli_compress_method = compressMethod;
+                c->ssl_cert_compression_brotli_decompress_method = decompressMethod;
+            }
+            break;
+        case TLSEXT_cert_compression_zstd:
+            result = SSL_CTX_add_cert_compression_alg(c->ctx, algorithmId,
+                direction & SSL_CERT_COMPRESSION_DIRECTION_COMPRESS ? zstd_compress_java : NULL,
+                direction & SSL_CERT_COMPRESSION_DIRECTION_DECOMPRESS ? zstd_decompress_java : NULL);
+            if (result) {
+                c->ssl_cert_compression_zstd_algorithm = algoRef;
```

Comment:
Same comment as above
```suggestion
                if (c->ssl_cert_compression_zstd_algorithm != NULL) {
                    (*e)->DeleteGlobalRef(e, c->ssl_cert_compression_zstd_algorithm);
                }
                c->ssl_cert_compression_zstd_algorithm = algoRef;
```